tests/conn_can/main.c

```diff
@@ -0,0 +1,658 @@
+            if (ret == 0) {
+                ret = conn_can_isotp_bind(&conn_isotp);
+            }
+            reply.type = msg.type;
+            reply.content.value = ret;
+            msg_reply(&msg, &reply);
+            break;
+        }
+        case CAN_MSG_RECV_ISOTP:
+        {
+            int ret = 0;
+            msg_t reply;
+            reply.type = msg.type;
+            reply.content.value = ret;
+            msg_reply(&msg, &reply);
+            while ((ret = conn_can_isotp_recv(&conn_isotp, isotp_buf[thread_nb],
+                                              ISOTP_BUF_SIZE, TEST_CONN_CAN_RECV_TIMEOUT))
+                   <= ISOTP_BUF_SIZE && ret >= 0) {
+                printf("%d: %-8s ISOTP [%d] ",
+                       thread_nb, raw_can_dev_find_by_ifnum(conn_isotp.ifnum), ret);
+                /*printf("%d: ISOTP [%d] ",
+                       thread_nb, ret);*/
+                for (int i = 0; i < ret; i++) {
+                    printf(" %02X", isotp_buf[thread_nb][i]);
+                }
+                printf("\n");
+            }
+            printf("%d: recv terminated: ret=%d\n", thread_nb, ret);
+            break;
+        }
+        case CAN_MSG_CLOSE_ISOTP:
+        {
+            msg_t reply;
+            int ret = conn_can_isotp_close(&conn_isotp);
+            reply.type = msg.type;
+            reply.content.value = ret;
+            msg_reply(&msg, &reply);
+            break;
+        }
+        case CAN_MSG_SEND_ISOTP:
+        {
+            msg_t reply;
+            can_opt_t *opt = msg.content.ptr;
+            int ret = conn_can_isotp_send(&conn_isotp, opt->data, opt->data_len, 0);
+            reply.type = msg.type;
+            reply.content.value = ret;
+            msg_reply(&msg, &reply);
+            break;
+        }
+#endif /* MODULE_CAN_ISOTP */
+        default:
+            printf("%d: _receive_thread: received unknown message\n", thread_nb);
+            break;
+        }
+    }
+
+    return NULL;
+}
+
+static const shell_command_t _commands[] = {
+    {"can", "CAN functions", _can_handler},
+    { NULL, NULL, NULL},
+};
+
+static void *_shell_thread(void *args)
+{
+    char line_buf[SHELL_DEFAULT_BUFSIZE];
+
+    (void) args;
+    msg_t msg_queue[8];
+    msg_init_queue(msg_queue, 8);
+
+    shell_run(_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
+
+    return NULL;
+}
+
+int main(void)
+{
+    thread_create(cmd_line_stack, THREAD_STACKSIZE, THREAD_PRIORITY_IDLE - 1,
+                  THREAD_CREATE_STACKTEST, _shell_thread, NULL, "cmd_line");
+
+    for (int i = 0; i < RCV_THREAD_NUMOF - 1; i++) {
+        receive_pid[i] = thread_create(thread_stack[i], THREAD_STACKSIZE,
+                                       THREAD_PRIORITY_MAIN + 1 + i,
+                                       THREAD_CREATE_STACKTEST, _receive_thread,
+                                       (void*)i, "receive_thread");
+    }
+
+
+    receive_pid[RCV_THREAD_NUMOF - 1] = sched_active_pid;
```

Comment:
When I started that app, I had some troubles with the thread priorities. I needed to reduce shell priority. I'm not sure if it applies anymore, though.